### PR TITLE
noto-fonts-lgc-plus: init at 20201206-phase3

### PR DESCRIPTION
**Pull request by laalsaas**
###### Description of changes
This is an effort to save users from installing every single noto-font on their system, making font selection in some programs almost unusable.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/data/fonts/noto-fonts/default.nix
Comment (SuperSandro2000) [inline pkgs/data/fonts/noto-fonts/default.nix:21]:
```suggestion
      version = "20201206-phase3";
```

File: pkgs/data/fonts/noto-fonts/default.nix
Comment (SuperSandro2000) [inline pkgs/data/fonts/noto-fonts/default.nix:46]:
```suggestion
```

File: pkgs/data/fonts/noto-fonts/default.nix
Comment (SuperSandro2000) [inline pkgs/data/fonts/noto-fonts/default.nix:126]:
```suggestion
    # allow the user to override which variants to 
    variants = if variants  == [] then [install
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

**Comment by Mathnerd314** [comment]
I guess this is fine, [FreeBSD](https://www.freshports.org/x11-fonts/noto-basic) does this, as have some [Manjaro users](https://www.reddit.com/r/ManjaroLinux/comments/i2war8/which_noto_fonts_are_really_required_by_the_system/) in search of smaller install sizes. I would call it `noto-fonts-latin` though to make clear that it only works for English-speaking users and a few other lucky alphabets. The fonts included aren't really the base of anything, it's just that the Latin alphabet is super common. Per [WP](https://en.wikipedia.org/wiki/Languages_used_on_the_Internet#Content_languages_for_websites) Latin is something like 90% of websites, while CJK is 4%, Arabic is 1%, and others less. 

The other possibility I'll mention for posterity is to installing all the fonts as before, but to use fontconfig magic to merge all the fonts into one font entry. Fedora does this with a [complex Lua script](https://src.fedoraproject.org/rpms/google-noto-fonts/blob/rawhide/f/google-noto-fonts.spec#_47). I think it works pretty well for the common case of specifying "Noto Sans" and having all symbols available, but I'm not sure it uses the correct font if a document specifies a specific font like "Noto Sans Sinhala". I don't know if this fits your requirements since it doesn't save any disk space.

**Comment by Mathnerd314** [comment]
Still another option is to write [a fontconfig to exclude all but the latin fonts](https://unix.stackexchange.com/questions/534284/reject-all-fonts-from-a-directory-except-one-with-fontconfig) and then load this for applications that don't have good font menus. But last I tried all the applications I use can handle the 200-ish fonts in Noto just fine. Which programs in particular are problematic?

**Comment by Mathnerd314** [comment]
> I think `noto-fonts-latin` is confusing *since* it also contains cyrillic and greek alphabet. what do you think about `noto-fonts-basic` like FreeBSD, or `noto-fonts-minimal`?

Alright, `noto-fonts-basic` or maybe `noto-fonts-common`.

>  splitting the fonts up in one font for each script is an upstream decision we should respect

The one-script-per-file format was basically chosen for convenience - Google only has to update one file when doing font updates, and it's  a 1-1 map with source files. Google has acknowledged this format is somewhat unwieldy for end users and there have been at least two Google-originated proposals to fix it (although neither has really been pushed by Google AFAICT):

* Merge the fonts into a few big groups (Historic, Southeast Asian, etc.), mentioned as a goal [here](https://github.com/notofonts/noto-fonts/issues/13) in 2013. Google implemented this in a [script](https://github.com/googlefonts/nototools/blob/main/nototools/merge_noto.py), which apparently is broken. But there is a [3rd party](https://github.com/satbyy/go-noto-universal) that has fixed it and provides 7 merged regional fonts.
* The fontconfig merge config method like I mentioned, this is discussed in an [open bug](https://github.com/notofonts/notobuilder/issues/13) that mentions providing such a fontconfig for distros. Admittedly no distro I know of uses such a fontconfig, but it was at least proposed by a (former) Google employee. 

> font programs (particulary, libfontconfig) can't handle fonts with a large number of symbols 

Yeah, there is a hard limit of 64K symbols in the font format, that Noto exceeds by like 5x (CJK alone is up at the limit). But messing with fontconfig doesn't create a font with an absurd number of symbols. Rather, it creates a lot of fonts with the same name and different version numbers ([FONTVERSION](https://lists.freedesktop.org/archives/fontconfig/2022-January/006860.html) allows specifying the sort order of identical fonts). Most programs only display one entry for all these seemingly-identical fonts, and the font fallback works properly to display all the characters. 

> Inkscape

Well, there is an [open Inkscape bug](https://gitlab.com/inkscape/inbox/-/issues/7797) to make font choosing with Noto easier. Other projects have made headway - LibreOffice has [proposed](https://design.blog.documentfoundation.org/2018/02/18/improvements-font-listing/) various designs (but not implemented any AFAICT), and GTK4 has implemented a [search by language](https://blog.gtk.org/2022/09/19/inside-the-gtk-font-chooser/) feature that filters out the majority of Noto fonts. So this problem could solve itself (or is arguably a minor annoyance at best).

**Comment by laalsaas** [comment]
> Alright, `noto-fonts-basic` or maybe `noto-fonts-common`.

I like `noto-fonts-common` and will edit this PR.

> But messing with fontconfig doesn't create a font with an absurd number of symbols. Rather, it creates a lot of fonts with the same name and different version numbers ([FONTVERSION](https://lists.freedesktop.org/archives/fontconfig/2022-January/006860.html) allows specifying the sort order of identical fonts). Most programs only display one entry for all these seemingly-identical fonts, and the font fallback works properly to display all the characters.

I apologize if my following statement is uninformed since i don't really know much about (lib)fontconfig, but this sounds cursed and not like something that should be shipped to any users except those who explicitely ask for it.

> > Inkscape
> 
> Well, there is an [open Inkscape bug](https://gitlab.com/inkscape/inbox/-/issues/7797) to make font choosing with Noto easier.

I know, said friend of mine has opened this bug.

Overall, your comment mentions lots of interessting facts and approaches, but I can't conclude from it what's your opinion on this PR. I think that this PR is currently the easiest way of making font selection less annoying for every plasma user. I think that it should be merged (but I'm obviously biased on this, since I *opened* this PR) as is (except for the name change discussed above), and then reducing the clutter for the main package (i.e. by merging the fonts as you suggested), should be part of a seperate effort. But maybe that „effort“ is also waiting for google to provide a solution, which would have the advantage of uniformity, however looking at the age of the comments, I doubt this will happen anytime soon.

Could you agree with this strategy or do you think this PR should be drastically changed to implement one of the approaches you mentioned?

**Comment by Mathnerd314** [comment]
> Could you agree with this strategy or do you think this PR should be drastically changed to implement one of the approaches you mentioned?

Well, my preferred setup is one where full Unicode coverage is provided. Noto was designed for "no tofu" and I don't want a situation where [List of Egyptian hieroglyphs](https://en.wikipedia.org/wiki/List_of_Egyptian_hieroglyphs) doesn't display properly or a friend can't borrow my computer to load his foreign-language website. So ideally a solution would preserve Unicode coverage. The fontconfig unification trick is indeed a bit cursed but Fedora has shipped it for the Droid fonts for years. So ideally you would work on unifying Noto the same way.

But I'm not going to force you. If you want to get this PR merged with the current approach I would say it's not impossible. NixOS is designed to be customizable and using a subset of Noto seems like a decent configuration. I could see it being used on some of the virtual KDE smoke tests for example, for a smaller closure size.

The main sticking point I see is changing the Plasma default font. Probably this will disrupt many users. @ttuegel is in charge of it, maybe he can weigh in. I would prefer to see a configuration switch which allows removing the default fonts or changing between noto-fonts-common and noto-fonts.

**Comment by laalsaas** [comment]
> > Could you agree with this strategy or do you think this PR should be drastically changed to implement one of the approaches you mentioned?
> 
> Well, my preferred setup is one where full Unicode coverage is provided. Noto was designed for "no tofu". So ideally a solution would preserve Unicode coverage.

Well, full unicode coverage is not affected for the existing `noto-fonts` package. (Though it admitedly is for those who only have installed Noto by using the plasma module)

> The main sticking point I see is changing the Plasma default font. Probably this will disrupt many users. @ttuegel is in charge of it, maybe he can weigh in. I would prefer to see a configuration switch which allows removing the default fonts or changing between noto-fonts-common and noto-fonts.

I'd argue that installing only `noto-fonts-common` would be the more sane default, but i can clearly see that it's problematic to change defaults towards removing functionality, esspecially when these defaults are already in use by thousands of users, who never bothered to look what's going on under the hood. The easiest approach would be to expose an option `plasma5.notoFontPackage`, which by default is set to `pkgs.noto-font`, but the user can easily set this to `pkgs.noto-fonts-common`. This would be relatively trivial, but I doubt that it would get widespread use.

Here is another, a little more tricky approach which I'm currently working on: Detecting the locale of the user, as well as all of his `i18n.supportedLocales`, and then installing exactly the fonts needed for these locales (plus maybe those already listed in `noto-fonts-common`, to ensure you can display latin letters, which is essential for non-translated programs, as well as nix code). i.e. Noto Sans Arabic for ar_AE, and so on. What do you think about making use of this mechanism in the plasma module? I think that it would disrupt least users, and those who really want full unicode coverage can still get it by putting `noto-fonts` in their `fonts.fonts`. However, this will take me a while, since I'll have to examine all the noto fonts in regard of which unix-locale they fit into.

**Comment by laalsaas** [comment]
> Here is another, a little more tricky approach which I'm currently working on: Detecting the locale of the user, as well as all of their `i18n.supportedLocales`, and then installing exactly the fonts needed for these locales (plus maybe those already listed in `noto-fonts-common`, to ensure you can display latin letters, which is essential for non-translated programs, as well as nix code). i.e. Noto Sans Arabic for ar_AE, and so on. What do you think about making use of this mechanism in the plasma module? I think that it would disrupt least users, and those who really want full unicode coverage can still get it by putting `noto-fonts` in their `fonts.fonts`. However, this will take me a while, since I'll have to examine all the noto fonts in regard of which unix-locale they fit into.

I am going to implement the second aproach, but it's going to take me some time

**Comment by Mathnerd314** [comment]
Sorry for late response, been sick the past week and of course it was Thanksgiving.

I think you should add a noto-fonts-uncommon package (or better name), so that the packages are a partition of the fonts like so:
```
                        weights

                     KDE      others
                   ┌─────────┬──────┐
           Latin   │common   │      │
           Greek   │         │      │
           Cyrillic│         │      │
languages          ├─────────┤extra │
           Hebrew  │         │      │
           Tamil   │uncommon │      │
           190 more│         │      │
                   └─────────┴──────┘
```


I think the font detection thing sounds a bit suspicious as the idea of NixOS is to have declarative configuration and magically detecting fonts goes against that. At best you could do a switch for plasma, which defaults to some logic based on `i18n.defaultLocale`  / `i18n.supportedLocales`, and use that to decide if -uncommon should be installed. I think doing per-locale packages would be overkill.

**Comment by emilazy** [comment]
For what it's worth I think breaking support for non-European languages out of the box would be a significant regression compared to other distributions and operating systems. No objection to having a smaller package available opt-in, of course, or addressing the UI issues in another way.

(As far as naming goes, I think that a glance at a list of the most widely spoken languages would be a good idea before declaring e.g. Greek as "common" and Hindi as "uncommon".)

**Comment by Mathnerd314** [comment]
> (As far as naming goes, I think that a glance at a list of the most widely spoken languages would be a good idea before declaring e.g. Greek as "common" and Hindi as "uncommon".)

That was up [here](https://github.com/NixOS/nixpkgs/pull/201951#issuecomment-1324328964). English is >60% of websites and Russian (Cyrillic) is second at 5.4%. Greek isn't really present as a spoken language but Greek symbols show up in a lot of math equations. So the single Noto font covers quite a chunk of the web (>90%). I don't know why spoken languages would be relevant, it seems Chinese/Hindi/Arabic speakers mainly use English online (probably because of terrible input methods).

**Comment by emilazy** [comment]
I can't speak for Hindi (I picked it at random) but as far as Chinese goes that's completely untrue... a quick glance at Weibo (which is comparably popular to Twitter) or similar will show huge amounts of Chinese-only content. In fact many large English sites are blocked outright. I'm somewhat skeptical of those Wikipedia figures; they're sourced from W3Techs who don't publish enough about their methodology for me to really pick at it but what they do say seems like it could potentially introduce systemic bias. But anyway I don't see why the package naming should necessarily be based on the frequency of the languages on the web, especially based on one Western source that primarily specializes in measuring server technology.

The names don't really matter much though; I mostly just think that we should not ship a worse multilingual experience out of the box than e.g. Fedora, Ubuntu, macOS or Windows, and any UI issues should be addressed as UI issues. Merging the fonts into a single UI entry seems fine as long as it doesn't break websites and applications that request the language-specific names (perhaps those could be included as separate but hidden entries, if that's necessary and viable), and offering a slimmed down package is a good idea for people who want or need it, but I don't think gutting language support out of the box is a good approach.

**Comment by betalars** [comment]
> I think the font detection thing sounds a bit suspicious as the idea of NixOS is to have declarative configuration and magically detecting fonts goes against that. At best you could do a switch for plasma, which defaults to some logic based on `i18n.defaultLocale` / `i18n.supportedLocales`, and use that to decide if -uncommon should be installed. I think doing per-locale packages would be overkill.

I don't like the idea, that users need to install the uncommon font package to be able to automatically use the fonts, that are most common for them, when we can make a mostly good enough guess based on a users locale.

Programs are choosing reasonable presets based on a users Locale all the time. Including on NixOS. I really do not see, why we would not consider a users locale in determining, what fonts are common to them.

**Comment by betalars** [comment]
Also: having some "black-magic" options, that automagically fix issues with software weirdness on nixos is neither a bad nor an unusual thing if you ask me.

see:
- `programs.steam.enable = true;`
- `programs.ausweisapp.openFirewall`
for instance

**Comment by SuperSandro2000** [comment]
> At best you could do a switch for plasma, which defaults to some logic based on `i18n.defaultLocale` / `i18n.supportedLocales`, and use that to decide if -uncommon should be installed. I think doing per-locale packages would be overkill.

i18n has nothing to do with different languages the browser could display and selecting fonts based on that does not make much sense.

**Comment by Mathnerd314** [comment]
> a quick glance at Weibo (which is comparably popular to Twitter) or similar will show huge amounts of Chinese-only content.

Alright, I guess we should use the [internet users by language](https://en.wikipedia.org/wiki/Languages_used_on_the_Internet#Internet_users_by_language)  numbers. That has English at 25.9% followed by Chinese at 19.4%. The formula is simply total number of speakers * percent population that uses the internet, so should include social media etc.

The naming `common`/`uncommon` was just what laalsaas liked, if you have a better scheme then we can use that. [SE](https://linguistics.stackexchange.com/questions/43100/is-there-a-collective-term-for-the-latin-cyrillic-and-greek-alphabets) suggests the terms Euroscript and LGC for the Latin-Greek-Cyrillic. For the other scripts there is not a good collective name. Probably we want a `-current`/`-historic` split, because nobody speaks the historic languages hence the historic languages actually can be gutted without offending anyone. Although CJK is still its own font.

Packaging sans and serif together like they are now is easiest, because there are some misc fonts that are not part of the main Noto Sans and Noto Serif families, discussed in [the FAQ](https://fonts.google.com/noto/use#faq). But they can be roughly categorized as sans or serif :

* the UI fonts have sans/serif in the name
* Noto Sans Adlam Unjoined, Noto Kufi Arabic, Noto Looped Lao, Noto Looped Thai, Noto Traditional Nushu are sans-ish
* Noto Naskh Arabic, Noto Nastaliq Urdu, Noto Rashi Hebrew are serif-ish

The easy option is to just pretend these are sans/serif. But properly speaking they should be split up into yet another package, because only Noto Sans is needed for KDE. I'll just ignore idea that for now. Then the packages would be `noto-sans-lgc`, `noto-sans-cjk`, `noto-sans-current`, `noto-sans-historic`, and `noto-sans-extra`, x2 for serif, plus the emoji font. Phew.

> I mostly just think that we should not ship a worse multilingual experience out of the box than e.g. Fedora, Ubuntu, macOS or Windows, and any UI issues should be addressed as UI issues.

Indeed, I think the default font set is where laalsaas gets outvoted. I still haven't merged my [`enableUnicodeCoverage` font switch](https://github.com/NixOS/nixpkgs/commit/c32be2f4190702dbbfb69fb71b328509e7e183a4) but defaulting that to true on desktop installs seems like a reasonable configuration. Then people who want to mess with slimmed down fonts and potential tofu can turn it off.

> Merging the fonts into a single UI entry seems fine as long as it doesn't break websites and applications that request the language-specific names (perhaps those could be included as separate but hidden entries, if that's necessary and viable), 

I just don't know fontconfig well enough to say if it's a viable solution. The Fedora maintainers say it's possible and should work great but they haven't actually done it for Noto. I guess I have to burn a weekend messing with my [Python script](https://github.com/NixOS/nixpkgs/blob/2e65d949908f3df63d7abfa07f9101f33971176c/pkgs/data/fonts/notofonts/analyze_scan.py) again to generate a fontconfig and see if it works.

Note that because there are the variant fonts the font menu would still have ~20 Noto entries.

**Comment by emilazy** [comment]
> I still haven't merged my [`enableUnicodeCoverage` font switch](https://github.com/NixOS/nixpkgs/commit/c32be2f4190702dbbfb69fb71b328509e7e183a4) but defaulting that to true on desktop installs seems like a reasonable configuration.

For what it's worth, I also went on a spree of trying to get as complete Unicode coverage as possible, and found that enabling `hanazono` and `babelstone-han` had a not insignificant performance impact; things chugged a bit when rendering and scrolling tests that used a wide range of obscure CJK glyphs. So there is *some* amount of tradeoff to be made here, although I'm not sure how much effect on performance it had when only more common glyphs are used (I can imagine the font selection code not being especially well-optimized for it).

**Comment by Mathnerd314** [comment]
At this point I think the main issue with 100% unicode coverage is CJK - Noto / unifont cover the rest. hanazono has CJK Unified Ideographs Extensions A-F, but has not been updated and Unicode 15 is up to Extension H. But from the [table](https://www.babelstone.co.uk/Fonts/Han.html) babelstone-han doesn't have anything close to 100%. It looks like [plangothik](https://github.com/Fitzgerald-Porthmouth-Koenigsegg/Plangothic/blob/main/README.en.md) provides G/H under the SIL OFL font license, but the readme also says "It is strictly prohibited for any individual, enterprise or team to sell the font file (including but not limited to [...] bundling it with other fonts)", which the OFL license [allows](https://scripts.sil.org/cms/scripts/page.php?item_id=OFL-FAQ_web#98ed3755), so maybe the license is not actually OFL.

**Comment by laalsaas** [comment]
Ok, so let me suggest a compromise:
- create the `noto-fonts-common` package (or any other name if there are objections)
- add a new option to the plasma module: `notoPackage`, which defaults to `pkgs.noto-fonts`, but the user can set it, i.e. `pkgs.noto-fonts-common`.

I don't really see a value in adding a -uncommon package. If you want all Noto fonts, you can just install the regular noto-fonts package, and I don't see a use case where users want to install all fonts except these in -common.

Could you agree to this?

**Comment by Mathnerd314** [comment]
> I don't really see a value in adding a -uncommon package. If you want all Noto fonts, you can just install the regular noto-fonts package, and I don't see a use case where users want to install all fonts except these in -common.

The advantage of -uncommon is that the packages form a partition. So you either install common alone or common+uncommon and there are no duplicated fonts. With noto-fonts and noto-fonts-common packages there would be [around](https://freebsd.pkgs.org/13/freebsd-amd64/noto-basic-2.0_1.pkg.html) 1.26 MB compressed / 4.40 MB uncompressed of duplicated fonts. It's similar to how -extra does not include the main package, avoiding 62 megabytes of duplication.

**Comment by emilazy** [comment]
The source repositories for the non-language-specific "Noto Sans" fonts is called [notofonts/latin-greek-cyrillic](https://github.com/notofonts/latin-greek-cyrillic), so I would potentially suggest `noto-fonts-latin-greek-cyrillic`. Though in that case including Math, Music and Symbols is arguably confusing, but maybe omitting them would be for the best in a variant designed to minimize disk space and font picker entries. (I suppose we could have packages for every single font if we really wanted...)


I don't see the point of an -uncommon package if the smaller package will only be installed by opt-in user choice; you should never end up with a situation where the complete package and the smaller one are both installed unless you explicitly set it up that way.

**Comment by laalsaas** [comment]
> Though in that case including Math, Music and Symbols is arguably confusing, but maybe omitting them would be for the best

I think the usefullness of Music is debatable, but Math is a must-have. What do you think about `noto-fonts-lgc-plus` or `noto-fonts-european`?

**Comment by emilazy** [comment]
Sure, this seems fine to me. Minor nits:

* The `variants` parameter for `noto-fonts-lgc-plus` seems strange to me because if you do override it it's going to be a meaningfully different package with the same name; probably just exposing mkNoto is good enough?
* Given that the abbreviated name is somewhat obscure, it'd be good to override `meta.description` with an explanation of what it contains.
* Perhaps instead of an option in the KDE to override the Noto package, just have an option to disable adding fonts entirely? Then you can add whatever UI and monospace font you want to `fonts.fonts` yourself. Unless Plasma has a hard dependency on Noto Sans and Hack?

**Review by Mathnerd314 — APPROVED** [review]
Looks good to me.

**Comment by laalsaas** [comment]
@emilazy I implemented your first two points. regarding the last one: kde wants noto sans. It's their default font. You could probably still use the programs without it, but I think it's reasonable to keep it as it is now. See: https://develop.kde.org/hig/style/typography/

File: pkgs/data/fonts/noto-fonts/default.nix
Comment (SuperSandro2000) [inline pkgs/data/fonts/noto-fonts/default.nix:171]:
```suggestion
      emojiPythonEnv = buildPackages.python3.withPackages (p: with p; [ fonttools nototools ]);
```

**Review by SuperSandro2000 — APPROVED** [review]

**Pull request merged**
